### PR TITLE
Reset Go settings for protoc dependencies

### PR DESCRIPTION
**Pull request by fmeum**
**What type of PR is this?**

Bug fix

**What does this PR do? Why is it needed?**

Reduces the number of essentially distinct configurations in which `protoc` as a dependency of `go_proto_library` is analyzed. 

Note that this may not (yet) result in a significant reduction of *rebuilds* since Bazel is not particular friendly to reset transitions at the moment: It encodes some information about the path of transitions taken into the build configuration, which e.g. prevents transitions from ever returning to the top-level target configuration. But this will be solved by https://github.com/bazelbuild/bazel/issues/14023, which is very actively being worked on. In the meantime, this change should at least not regress the current behavior further.

**Which issues(s) does this PR fix?**

Fixes a part of #2999.

**Other notes for review**

@robfig If you think that this PR is useful, I would also look into a follow-up that resets the non-Go attrs, e.g. `data` and `embedsrcs`, of rules such as `go_library` to their values before `go_transition`. Ideally, `rules_go` settings set by transitions would only ever propagate to rules that provide a Go provider. Since this requires memoizing the state of settings before a transition, this will be a bit more complicated, which is why I would like to get your feedback on this simpler case first.

### CODE REVIEW

**Comment by robfig** [comment]
Hi @fmeum , thanks for the contribution. We would definitely like to avoid cache thrashing. I don't personally have any knowledge about transitions, but @blico volunteered to study up and review. Apologies for the delay, but be assured we haven't forgotten and would like to get this merged.

**Comment by fmeum** [comment]
No worries, let me know if you need any additional explanations.

**Comment by fmeum** [comment]
I rebased the PR onto master after my other changes were merged.

**Comment by fmeum** [comment]
@blico Friendly ping as the required commit has landed in Bazel: https://github.com/bazelbuild/bazel/commit/52d1d4ae9bfee7379093d46258fa3a56e4d69e61

**Comment by fmeum** [comment]
@robfig I just submitted https://github.com/bazelbuild/rules_go/pull/3108, which is based on this PR and finished the leftover work. Along the way, it fixes a rules_go correctness bug reported in https://github.com/bazelbuild/bazel/issues/14626. Do you think one of the maintainers will have time to review either of those PRs in the near future?

**Review by achew22 — APPROVED** [review]

File: go/private/rules/transition.bzl
Comment (linzhp) [inline go/private/rules/transition.bzl:268]: `_non_go_exec_reset_transition_keys` can be constructed from `_reset_transition_dict` too, because it as the same set of keys as `_non_go_exec_reset_transition_dict`. If we do that, then we can don't need `_non_go_exec_reset_transition_dict`, and instead setting `@io_bazel_rules_go//go/private:bootstrap_nogo` in   `_go_non_go_exec_reset_tool_transition_impl`

File: proto/BUILD.bazel
Comment (linzhp) [inline proto/BUILD.bazel:120]:
```suggestion
    name = "protoc",
```
maybe just name it "protoc"?

File: go/private/rules/transition.bzl
Comment (linzhp) [inline go/private/rules/transition.bzl:364]: This doc seems wrong

File: go/private/rules/transition.bzl
Comment (linzhp) [inline go/private/rules/transition.bzl:327]: Maybe keep this `go_reset_target` and name another one `non_go_reset_target`?

**Review by linzhp — CHANGES_REQUESTED** [review]
Passed the tests in Uber's Go monorepo. Left a few stylish comments. Looks good overall. Sorry for taking so long to review.

File: go/private/rules/transition.bzl
Comment (fmeum) [reply go/private/rules/transition.bzl]: I took the chance to improve the docs, but am not sure I understood this particular point.

File: go/private/rules/transition.bzl
Comment (fmeum) [reply go/private/rules/transition.bzl]: I ended up removing `_non_go_exec_reset_transition_keys` as well in favor of `_reset_transition_keys`.

**Pull request merged**
